documentation/projects/proposals/python_packaging/20240412-implementation_plan_switching_to_a_new_python_package_manager.md

```diff
@@ -0,0 +1,207 @@
+
+  This is somewhat subjective, but the package manager should be fairly easy to
+  use, should not have too much magic, and should make Python workflows like
+  building Docker images and publishing packages to PyPI easier.
+
+- **comply with standards**
+
+  The package manager should not reinvent the wheel and should follow the
+  standards like PEP 621, PEP 517 and PEP 508. This is important for future
+  integrations and compatibility with other tools.
+
+- **be around for the long term**
+
+  The package manager should be actively developed and maintained, and should
+  have a good community around it. This is important for long-term support and
+  active maintenance in the future.
+
+## Decision making
+
+To decide the new package manager, many options were considered. Here are the
+options, what benefits they provided and why they were ultimately not chosen as
+the path forward.
+
+![](./star_history.png)
+
+### [Rye](https://rye-up.com/)
+
+#### Why?
+
+- all-in-one tool (at least trying to be)
+- provisions
+  [@indygreg's python-build-standalone](https://github.com/indygreg/python-build-standalone)
+- fastest among all options considered (written in Rust, uses
+  [`uv`](https://github.com/astral-sh/uv/) under the hood)
+- lockfile is `requirements.txt`
+- supports monorepo workspaces
+- developed by [Armin Ronacher](https://github.com/mitsuhiko)
+- fastest growing in terms of GitHub stars
+- supports multiple build backends via
+  [PyPA/build](https://build.pypa.io/en/stable/)
+
+#### Why not?
+
+- PEP 621 non-compliant
+- looks like
+  [1 core maintainer](https://github.com/astral-sh/rye/graphs/contributors)
+- [no stable v1 yet](https://github.com/astral-sh/rye/releases)
+- [considered "experimental"](https://github.com/search?q=repo:astral-sh/rye%20experimental&type=code)
+- [VC-backed](https://astral.sh/about)
+- [not supported by Renovate](https://github.com/renovatebot/renovate/issues/25273)
+  [⚠️ deal-breaker]
+- does not generate platform-independent lockfiles
+
+### [PDM](https://pdm-project.org/)
+
+#### Why?
+
+- PEP 621 compliant
+- supported by [Renovate](https://docs.renovatebot.com/modules/manager/pep621/)
+  and
+  [Dependabot](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/about-dependabot-version-updates#pip-and-pip-compile)
+- supports multiple build backends
+- supports monorepo workspaces
+- provisions
+  [@indygreg's python-build-standalone](https://github.com/indygreg/python-build-standalone)
+- [endorsed](https://github.com/astral-sh/rye/discussions/6#discussioncomment-5700997)
+  by creator of Rye!
+
+#### Why not?
+
+- looks like
+  [1 core maintainer](https://github.com/pdm-project/pdm/graphs/contributors)
+- fewest GitHub stars and forks indicative of low popularity
+- [flagship feature rejected](https://pdm-project.org/en/latest/usage/pep582/)
+- [does not support editable main dependencies](https://pdm-project.org/en/latest/usage/dependency/#editable-dependencies)
```

Comment:
Are editable dependencies relevant outside a development context? You can't use them to build, you need to use a copied version.

The way this works is, you'd specify the editable package in dev dependencies and a non-editable version in the regular dependencies. That way, when you install in dev mode, the editable version is referenced, when you install in non-dev mode, the real copied package is used.

I've tried this locally and it works fine. Have I missed something we need from editable packages? I don't think we even _want_ editable packages outside a dev context.

Reply:
Do you mean this?

```toml
[project]
name = "api"

dependencies = [
  "ov-attribution @ file:///${PROJECT_ROOT}/../packages/ov-attribution",
]

[tool.pdm.dev-dependencies]
sub_pkgs = [
  "-e ov-attribution @ file:///${PROJECT_ROOT}/../packages/ov-attribution"
]
```

I tried locking a project with such a setup but `pdm.lock` only says `editable=true`.
<img width="642" alt="Screenshot 2024-04-17 at 1 24 55 AM" src="https://github.com/WordPress/openverse/assets/16580576/7149eb93-c4de-48cd-a554-3105ddaa63c3">

Is that expected?

Reply:
I thought so. My assumption _was_ that the lock file would indicate it as editable, but then only install it as such when you selected the dev dependencies with `pdm install --dev`. However, when I unselect the dev groun locally with `pdm install --prod`, it still installs it as editable.

_But_, I can pass `--no-editable`, and it works fine. `pdm install --prod --no-editable` copies the code directly into the venv, rather than creating a `.pth` file. The important bit being that we can install it editable in dev so that we can e.g., map the directory in docker compose to the relative location referenced.

For a future prod-only dependencies version of the `api` target in the API Dockerfile, we'd use `pdm install --prod --no-editable --frozen-lockfile`.

I wonder if the best way to do this would actually be to install the prod dependencies only in the Dockerfile, then add an entrypoint to the compose service that runs `pdm install --dev` to bootstrap the development environment in the container for local use.

Reply:
@sarayourfriend I think @krysal has approached that before to resolve #1008. Maybe something to reconsider after this transition.